subsys/net/lib/sockets/sockets.c

```diff
@@ -168,3 +178,67 @@ ssize_t zsock_send(int sock, const void *buf, size_t len, int flags)
 
 	return len;
 }
+
+ssize_t zsock_recv(int sock, void *buf, size_t max_len, int flags)
+{
+	(void)flags;
+	struct net_context *ctx = INT_TO_POINTER(sock);
+	enum net_sock_type sock_type = net_context_get_type(ctx);
+	size_t recv_len = 0;
+
+	if (sock_type == SOCK_DGRAM) {
+		__ASSERT(0, "DGRAM is not yet handled");
+	} else if (sock_type == SOCK_STREAM) {
+		do {
```

Comment:
GCC does tail-call optimization, so there's no need to worry about increased stack usage. Essentially, if a function does `return function(...)`, instead of generating a CALL instruction (or their equivalent), it'll generate a JMP instruction (or their equivalent) after setting up the parameters.  It's very likely that there won't be any special code for the parameters, as it might be able to reuse the ones in the current call frame.